File: benches/arith_bench.rs
#![feature(test)]
extern crate test;
use test::Bencher;

#[macro_use]
extern crate ndarray;
extern crate num_ru;
extern crate ndarray_parallel;

use ndarray::*;
use num_ru::math::arithmetic::*;
use num_ru::math::sumproddif::*;

#[bench]
fn negative_bench_mid(b: &mut Bencher) {
    b.iter(|| {
        let input_arr = Array::from_elem((50, 50, 50), 1.0);
        let expected_arr = Array::from_elem((50, 50, 50), -1.0);
        let res_arr = input_arr.negative();
        assert_eq!(expected_arr, res_arr);
    });
}

#[bench]
fn negative_bench_large(b: &mut Bencher) {
    b.iter(|| {
        let input_arr = Array::from_elem((50, 50, 50, 50), 1.0);
        let expected_arr = Array::from_elem((50, 50, 50, 50), -1.0);
        let res_arr = input_arr.negative();
        assert_eq!(expected_arr, res_arr);
    });
}

#[bench]
fn sum_bench_mid(b: &mut Bencher) {
    b.iter(|| {
        let input_arr = Array::from_elem((50, 50, 50), 1.0);
        let res = input_arr.sum();
        assert_eq!(res, 125000.0);
    });
}

#[bench]
fn sum_bench_large(b: &mut Bencher) {
    b.iter(|| {
        let input_arr = Array::from_elem((50, 50, 50, 50), 1.0);
        let res = input_arr.sum();
        assert_eq!(res, 6250000.0);
    });
}

#[bench]
fn prod_bench_mid(b: &mut Bencher) {
    b.iter(|| {
        let input_arr = Array::from_elem((50, 50, 50), 1.0);
        let res = input_arr.prod();
        assert_eq!(res, 1.0);
    });
}

#[bench]
fn prod_bench_large(b: &mut Bencher) {
    b.iter(|| {
        let input_arr = Array::from_elem((50, 50, 50, 50), 1.0);
        let res = input_arr.prod();
        assert_eq!(res, 1.0);
    });
}

#[bench]
fn negative_bench_mid_rayon(b: &mut Bencher) {
    b.iter(|| {
        let input_arr = ArcArray::from_elem((50, 50, 50), 1.0);
        let expected_arr = ArcArray::from_elem((50, 50, 50), -1.0);
        let res_arr = input_arr.negative();
        assert_eq!(expected_arr, res_arr);
    });
}

#[bench]
fn negative_bench_large_rayon(b: &mut Bencher) {
    b.iter(|| {
        let input_arr = ArcArray::from_elem((50, 50, 50, 50), 1.0);
        let expected_arr = ArcArray::from_elem((50, 50, 50, 50), -1.0);
        let res_arr = input_arr.negative();
        assert_eq!(expected_arr, res_arr);
    });
}

#[bench]
fn sum_bench_large_rayon(b: &mut Bencher) {
    b.iter(|| {
        let input_arr = ArcArray::from_elem((50, 50, 50, 50), 1.0);
        let res = input_arr.sum();
        assert_eq!(res, 6250000.0);
    });
}

#[bench]
fn sum_bench_mid_rayon(b: &mut Bencher) {
    b.iter(|| {
        let input_arr = ArcArray::from_elem((50, 50, 50), 1.0);
        let res = input_arr.sum();
        assert_eq!(res, 125000.0);
    });
}

#[bench]
fn prod_bench_mid_rayon(b: &mut Bencher) {
    b.iter(|| {
        let input_arr = ArcArray::from_elem((50, 50, 50), 1.0);
        let res = input_arr.prod();
        assert_eq!(res, 1.0);
    });
}

#[bench]
fn prod_bench_large_rayon(b: &mut Bencher) {
    b.iter(|| {
        let input_arr = ArcArray::from_elem((50, 50, 50, 50), 1.0);
        let res = input_arr.prod();
        assert_eq!(res, 1.0);
    });
}

File: python_benches/arith_bench.py
import numpy as np
import timeit

def negative_bench_mid():
    input_arr = np.full((50, 50, 50), 1.0)
    expected_arr = np.full((50, 50, 50), -1.0)
    assert np.array_equal(np.negative(input_arr), expected_arr)
    return

def negative_bench_large():
    input_arr = np.full((50, 50, 50, 50), 1.0)
    expected_arr = np.full((50, 50, 50, 50), -1.0)
    assert np.array_equal(np.negative(input_arr), expected_arr)
    return

def sum_bench_mid():
    input_arr = np.full((50, 50, 50), 1.0)
    assert np.sum(input_arr) == 125000.0
    return

def prod_bench_mid():
    input_arr = np.full((50, 50, 50), 1.0)
    assert np.prod(input_arr) == 1.0
    return

def prod_bench_large():
    input_arr = np.full((50, 50, 50, 50), 1.0)
    assert np.prod(input_arr) == 1.0
    return

def bench(function_name):
    setup = "from __main__ import " + function_name
    res_str = "test: "
    res_str += '{:<15}'.format(function_name) + "  ... bench: "
    res_str += '{:^,}'.format(timeit.timeit(function_name + "()", setup=setup, number=500) / 500 * 1000000000)
    res_str += " ns/iter"

    print res_str

def bench_arith():
    bench("negative_bench_mid")
    bench("negative_bench_large")
    bench("sum_bench_mid")
    bench("prod_bench_mid")
    bench("prod_bench_large")

if __name__ == "__main__":
    print "Arithmetic Python NumPy Benchmarks"
    bench_arith()
    print "\n"
 
